cmds/console/console.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2015-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// console implements a basic console. It establishes a pair of files
+// to read from, the default being a UART at 0x3f8, but an alternative
+// being just stdin and stdout. It will also set up a root file system
+// using uroot.Rootfs, although this can be disabled as well.
+// Console uses a Go version of fork_pty to start up a shell, default
+// /ubin/rush. Console runs until the shell exits and then exits itself.
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+
+	"github.com/u-root/u-root/pkg/pty"
+	"github.com/u-root/u-root/uroot"
+)
+
+var (
+	serial    = flag.String("serial", "0x3f8", "use IO for stdin")
+	setupRoot = flag.Bool("setuproot", true, "Set up a root file system")
+)
+
+func main() {
+	fmt.Printf("console -- starting")
+	flag.Parse()
+
+	a := flag.Args()
+	if len(a) < 1 {
+		a = []string{"/ubin/rush"}
+	}
+
+	p, err := pty.New(a[0], a[1:]...)
+	if err != nil {
+		log.Fatalf("Can't open pty: %v", err)
+	}
+	// Make a good faith effort to set up root. This being
+	// a kind of init program, we do our best and keep going.
+	if *setupRoot {
+		uroot.Rootfs()
+	}
+
+	in, out := io.Reader(os.Stdin), io.Writer(os.Stdout)
+
+	if *serial != "" {
+		u, err := openUART(*serial)
+		if err != nil {
+			// can't happen but ... maybe.
+			fmt.Printf("Sorry, can't get a uart: %v", err)
+			os.Exit(1)
+		}
+		in, out = u, u
+	}
+
+	err = p.Start()
+	if err != nil {
```

Comment:
I actually decided not to embed in this case, it just didn't seem quite appropriate.